tensorflow/python/framework/python_op_gen.cc

```diff
@@ -355,18 +416,97 @@ string GenEagerPythonOp::Code() {
     return result_;
   }
 
   if (!AddEagerFallbackCode(parameters, output_sizes, num_outputs_expr,
-                            eager_not_allowed_error)) {
+                            eager_not_allowed_error, type_annotations)) {
     return result_;
   }
 
   return prelude_ + result_;
 }
 
+std::unordered_map<string, string> GenEagerPythonOp::GetTypeAnnotationMap() {
+  std::unordered_map<string, string> type_annotations;
+  // Mapping attrs to TypeVars
+  for (const auto& attr : op_def_.attr()) {
+    if (attr.type() == "type") {
+      const string type_var_name = "TV_" + op_def_.name() + "_" + attr.name();
+      type_annotations[attr.name()] = type_var_name;
+    } else if (attr.type() == "bool" || attr.type() == "float" ||
+               attr.type() == "int" || attr.type() == "bytes") {
+      type_annotations[attr.name()] = attr.type();
+    }
+  }
+
+  // Mapping input Tensors to their types
+  for (const auto& arg : op_def_.input_arg()) {
+    // Do not add type annotations to args that accept a sequence of Tensors
+    if (!arg.number_attr().empty()) continue;
+    type_annotations[arg.name()] = GetArgAnnotation(arg, type_annotations);
+  }
+
+  // Mapping output Tensor to its type
+  if (op_def_.output_arg_size() == 1) {
+    const auto& arg = op_def_.output_arg(0);
+    type_annotations[arg.name()] = GetArgAnnotation(arg, type_annotations);
+  }
+
+  return type_annotations;
+}
+
+// Generate TypeVars using attrs
+void GenEagerPythonOp::GenerateTypeVars(std::unordered_map<string, string>& type_annotations) {
+  bool added_typevar = false;
+  for (const auto& attr : op_def_.attr()) {
+    if (attr.type() == "type") {
+      std::vector<string> allowed_types;
+      for (int t : attr.allowed_values().list().type()) {
+        DataType dtype = static_cast<DataType>(t);
+        const string py_dtype = python_op_gen_internal::DataTypeToPython(dtype, "_dtypes.");
+        if (dtype_type.find(py_dtype) != dtype_type.end()) {
+          allowed_types.emplace_back(dtype_type[py_dtype]);
+        }
+      }
+
+      // If all dtypes are allowed, add them all
+      if (allowed_types.empty()) {
+        for (std::pair<string, string> map_dtype : dtype_type) {
+          allowed_types.emplace_back(map_dtype.second);
+        }
+      }
+
+      std::sort(allowed_types.begin(), allowed_types.end());
+
+      string typevar_dtypes;
+      for (std::vector<string>::iterator it = allowed_types.begin(); it != allowed_types.end(); ++it) {
+        if (!typevar_dtypes.empty()) strings::StrAppend(&typevar_dtypes, ", ");
+        strings::StrAppend(&typevar_dtypes, *it);
+      }
+
+      const string type_var_name = type_annotations[attr.name()];
+      strings::StrAppend(&result_, type_var_name, " = TypeVar(\"", type_var_name, "\", ", typevar_dtypes,")\n");
+      added_typevar = true;
+    }
+  }
+
+  if (added_typevar) strings::StrAppend(&result_, "\n");
+}
+
+// TODO(rahulkamat): Modify AddDefLine() to add return type annotation
```

Comment:
I think this todo is a bit out of context - it's better to move it next to the erase call, and explain that doing it would avoid having to erase from the string.